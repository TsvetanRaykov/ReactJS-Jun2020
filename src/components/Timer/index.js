import React, { Fragment, useState, useEffect } from 'react'
import { withRouter } from 'react-router-dom'
import { padZero } from '../../utils'

const QuizTimer = (props) => {
	const { timer, history } = props
	const { duration, start, complete, progress } = timer || {}
	const [minutes, setMinutes] = useState(0)
	const [seconds, setSeconds] = useState(0)

	const [timerHandler, setTimerHandler] = useState(0)

	history.listen(() => {
		clearTimeout(timerHandler)
	})

	const updateTime = (timer) => {
		setMinutes(parseInt(timer / 60, 10))
		setSeconds(parseInt(timer % 60, 10))
	}

	const startTimer = () => {
		const interval = 1000 // ms
		let timer = duration
		let expected = Date.now() + interval
		setTimerHandler(() => {
			return setTimeout(step, interval)
		})

		function step() {
			const dt = Date.now() - expected

			updateTime(timer)

			if (--timer < 0) {
				complete()
				clearTimeout(timerHandler)
				return
			}
			progress(timer)
			expected += interval
			setTimerHandler(() => {
				return setTimeout(step, Math.max(0, interval - dt))
			})
		}
	}

	useEffect(() => {
		updateTime(timer.duration)
	}, [timer])

	useEffect(() => {
		if (start) startTimer()
		else {
			clearTimeout(timerHandler)
		}
		return () => {
			clearTimeout(timerHandler)
		}
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [start])

	return <Fragment>{`${padZero(minutes)} : ${padZero(seconds)}`}</Fragment>
}

export default withRouter(QuizTimer)
