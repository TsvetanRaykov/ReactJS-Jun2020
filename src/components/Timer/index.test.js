import React from 'react'
import renderer from 'react-test-renderer'
import TestWrapper from '../../utils/test-wrapper'
import Timer from '.'

jest.mock('react', () => ({
	...jest.requireActual('react'),
	useLayoutEffect: jest.requireActual('react').useEffect,
}))

describe('<Timer />', () => {
	it('Should match the snapshot', () => {
		const wrapper = renderer.create(
			<TestWrapper>
				<Timer />
			</TestWrapper>
		)
		expect(wrapper).toMatchSnapshot()
	})
})
