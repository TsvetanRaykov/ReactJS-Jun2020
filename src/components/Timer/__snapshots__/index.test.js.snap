// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timer /> Should match the snapshot 1`] = `"00 : 00"`;
